// standardized-audio-context/build/es2015/factories/offline-audio-destination-node.js
import { OfflineAudioNodeProxy } from '../offline-audio-node';
export class OfflineAudioDestinationNodeFakerProxy extends OfflineAudioNodeProxy {
    constructor({ fakeNodeStore, offlineAudioContext }) {
        super({
            channelCount: 2,
            channelCountMode: 'max',
            channelInterpretation: 'speakers',
            fakeNodeStore,
            numberOfInputs: 1,
            numberOfOutputs: 0,
            offlineAudioContext
        });
    }
    get maxChannelCount() {
        // @todo
        return 2;
    }
}
export class OfflineAudioDestinationNodeFaker {
    constructor({ fakeNodeStore, offlineAudioContext }) {
        this._node = null;
        this._proxy = new OfflineAudioDestinationNodeFakerProxy({ fakeNodeStore, offlineAudioContext });
        this._sources = new Map();
        fakeNodeStore.set(this._proxy, this);
    }
    get proxy() {
        return this._proxy;
    }
    render(offlineAudioContext) {
        if (this._node !== null) {
            return Promise.resolve(this._node);
        }
        this._node = offlineAudioContext.destination;
        const promises = Array
            .from(this._sources)
            .map(([source, { input, output }]) => {
            // For some reason this currently needs to be a function body with a return statement. The shortcut syntax causes an error.
            return source
                .render(offlineAudioContext)
                .then((node) => node.connect(this._node, output, input));
        });
        return Promise
            .all(promises)
            .then(() => this._node);
    }
    wire(source, output, input) {
        this._sources.set(source, { input, output });
        return this._proxy;
    }
    unwire(source) {
        this._sources.delete(source);
    }
}
export class OfflineAudioDestinationNodeFakerFactory {
    create({ fakeNodeStore, offlineAudioContext }) {
        return new OfflineAudioDestinationNodeFaker({ fakeNodeStore, offlineAudioContext });
    }
}
