// standardized-audio-context/build/es2015/factories/offline-audio-buffer-source-node.js
import { OfflineAudioNodeProxy } from '../offline-audio-node';
export class OfflineAudioBufferSourceNodeFakerProxy extends OfflineAudioNodeProxy {
    constructor({ fakeNodeStore, offlineAudioContext }) {
        super({
            channelCountMode: 'max',
            channelInterpretation: 'speakers',
            fakeNodeStore,
            numberOfInputs: 0,
            numberOfOutputs: 1,
            offlineAudioContext
        });
        this._buffer = null;
        this._ownFakeNodeStore = fakeNodeStore;
    }
    get buffer() {
        return this._buffer;
    }
    set buffer(value) {
        // @todo Allow to set the buffer only onces.
        this._buffer = value;
    }
    get onended() {
        // @todo
        return null;
    }
    set onended(value) {
        // @todo
        value; // tslint:disable-line:no-unused-expression
    }
    get detune() {
        // @todo Fake a proper AudioParam.
        const audioParam = {
            cancelScheduledValues: (startTime) => {
                startTime; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            defaultValue: 0,
            exponentialRampToValueAtTime: (value, endTime) => {
                endTime; // tslint:disable-line:no-unused-expression
                value; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            linearRampToValueAtTime: (value, endTime) => {
                endTime; // tslint:disable-line:no-unused-expression
                value; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            setTargetAtTime: (target, startTime, timeConstant) => {
                startTime; // tslint:disable-line:no-unused-expression
                target; // tslint:disable-line:no-unused-expression
                timeConstant; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            setValueAtTime: (value, startTime) => {
                startTime; // tslint:disable-line:no-unused-expression
                value; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            setValueCurveAtTime: (values, startTime, duration) => {
                duration; // tslint:disable-line:no-unused-expression
                startTime; // tslint:disable-line:no-unused-expression
                values; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            value: 0
        };
        return audioParam;
    }
    get loop() {
        // @todo
        return false;
    }
    get loopEnd() {
        // @todo
        return 0;
    }
    get loopStart() {
        // @todo
        return 0;
    }
    get playbackRate() {
        // @todo Fake a proper AudioParam.
        const audioParam = {
            cancelScheduledValues: (startTime) => {
                startTime; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            defaultValue: 1,
            exponentialRampToValueAtTime: (value, endTime) => {
                endTime; // tslint:disable-line:no-unused-expression
                value; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            linearRampToValueAtTime: (value, endTime) => {
                endTime; // tslint:disable-line:no-unused-expression
                value; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            setTargetAtTime: (target, startTime, timeConstant) => {
                startTime; // tslint:disable-line:no-unused-expression
                target; // tslint:disable-line:no-unused-expression
                timeConstant; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            setValueAtTime: (value, startTime) => {
                startTime; // tslint:disable-line:no-unused-expression
                value; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            setValueCurveAtTime: (values, startTime, duration) => {
                duration; // tslint:disable-line:no-unused-expression
                startTime; // tslint:disable-line:no-unused-expression
                values; // tslint:disable-line:no-unused-expression
                return audioParam;
            },
            value: 1
        };
        return audioParam;
    }
    start(when = 0, offset = 0, duration) {
        const faker = this._ownFakeNodeStore.get(this);
        faker.start = { duration, offset, when };
    }
    stop(when = 0) {
        // @todo
        when; // tslint:disable-line:no-unused-expression
    }
}
export class OfflineAudioBufferSourceNodeFaker {
    constructor({ fakeNodeStore, offlineAudioContext }) {
        this._node = null;
        this._proxy = new OfflineAudioBufferSourceNodeFakerProxy({ fakeNodeStore, offlineAudioContext });
        this._sources = new Map();
        this._start = null;
        fakeNodeStore.set(this._proxy, this);
    }
    get proxy() {
        return this._proxy;
    }
    set start(value) {
        this._start = value;
    }
    render(offlineAudioContext) {
        if (this._node !== null) {
            return Promise.resolve(this._node);
        }
        this._node = offlineAudioContext.createBufferSource();
        this._node.buffer = this._proxy.buffer;
        if (this._start !== null) {
            const { duration, offset, when } = this._start;
            if (duration === undefined) {
                this._node.start(when, offset);
            }
            else {
                this._node.start(when, offset, duration);
            }
        }
        const promises = Array
            .from(this._sources)
            .map(([source, { input, output }]) => {
            // For some reason this currently needs to be a function body with a return statement. The shortcut syntax causes an error.
            return source
                .render(offlineAudioContext)
                .then((node) => node.connect(this._node, output, input));
        });
        return Promise
            .all(promises)
            .then(() => this._node);
    }
    wire(source, output, input) {
        this._sources.set(source, { input, output });
        return this._proxy;
    }
    unwire(source) {
        this._sources.delete(source);
    }
}
export class OfflineAudioBufferSourceNodeFakerFactory {
    create({ fakeNodeStore, offlineAudioContext }) {
        return new OfflineAudioBufferSourceNodeFaker({ fakeNodeStore, offlineAudioContext });
    }
}
