// angular-audio-context/build/esm/audio-context.module.js
import { CommonModule } from '@angular/common';
import { NgModule, OpaqueToken } from '@angular/core';
import { AudioContext as StandardizedAudioContextAudioContext, isSupported as standardizedAudioContextModuleIsSupported } from 'standardized-audio-context';
export var AudioContext = new OpaqueToken('AUDIO_CONTEXT_CONSTRUCTOR'); // tslint:disable-line:variable-name
// tslint:disable-line:variable-name
export function audioContextConstructorFactory() {
    return new StandardizedAudioContextAudioContext();
}
export var isSupported = new OpaqueToken('IS_SUPPORTED_PROMISE');
export function isSupportedFactory() {
    return standardizedAudioContextModuleIsSupported;
}
var AudioContextModule = (function () {
    function AudioContextModule() {
    }
    return AudioContextModule;
}());
export { AudioContextModule };
AudioContextModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                providers: [
                    { provide: AudioContext, useFactory: audioContextConstructorFactory },
                    { provide: isSupported, useFactory: isSupportedFactory }
                ]
            },] },
];
/** @nocollapse */
AudioContextModule.ctorParameters = function () { return []; };
