// standardized-audio-context/build/esm/factories/offline-biquad-filter-node.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { OfflineAudioNodeProxy } from '../offline-audio-node';
var OfflineBiquadFilterNodeFakerProxy = (function (_super) {
    __extends(OfflineBiquadFilterNodeFakerProxy, _super);
    function OfflineBiquadFilterNodeFakerProxy(_a) {
        var fakeNodeStore = _a.fakeNodeStore, nativeNode = _a.nativeNode, offlineAudioContext = _a.offlineAudioContext;
        var _this = _super.call(this, {
            channelCountMode: 'max',
            channelInterpretation: 'speakers',
            fakeNodeStore: fakeNodeStore,
            numberOfInputs: 1,
            numberOfOutputs: 1,
            offlineAudioContext: offlineAudioContext
        }) || this;
        _this._nativeNode = nativeNode;
        _this._type = nativeNode.type;
        return _this;
    }
    Object.defineProperty(OfflineBiquadFilterNodeFakerProxy.prototype, "detune", {
        get: function () {
            // @todo Fake a proper AudioParam.
            var audioParam = {
                cancelScheduledValues: function (startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                defaultValue: 0,
                exponentialRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                linearRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setTargetAtTime: function (target, startTime, timeConstant) {
                    startTime; // tslint:disable-line:no-unused-expression
                    target; // tslint:disable-line:no-unused-expression
                    timeConstant; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueAtTime: function (value, startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueCurveAtTime: function (values, startTime, duration) {
                    duration; // tslint:disable-line:no-unused-expression
                    startTime; // tslint:disable-line:no-unused-expression
                    values; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                value: 0
            };
            return audioParam;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfflineBiquadFilterNodeFakerProxy.prototype, "frequency", {
        get: function () {
            // @todo Fake a proper AudioParam.
            var audioParam = {
                cancelScheduledValues: function (startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                defaultValue: 350,
                exponentialRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                linearRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setTargetAtTime: function (target, startTime, timeConstant) {
                    startTime; // tslint:disable-line:no-unused-expression
                    target; // tslint:disable-line:no-unused-expression
                    timeConstant; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueAtTime: function (value, startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueCurveAtTime: function (values, startTime, duration) {
                    duration; // tslint:disable-line:no-unused-expression
                    startTime; // tslint:disable-line:no-unused-expression
                    values; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                value: 350
            };
            return audioParam;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfflineBiquadFilterNodeFakerProxy.prototype, "gain", {
        get: function () {
            // @todo Fake a proper AudioParam.
            var audioParam = {
                cancelScheduledValues: function (startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                defaultValue: 0,
                exponentialRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                linearRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setTargetAtTime: function (target, startTime, timeConstant) {
                    startTime; // tslint:disable-line:no-unused-expression
                    target; // tslint:disable-line:no-unused-expression
                    timeConstant; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueAtTime: function (value, startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueCurveAtTime: function (values, startTime, duration) {
                    duration; // tslint:disable-line:no-unused-expression
                    startTime; // tslint:disable-line:no-unused-expression
                    values; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                value: 0
            };
            return audioParam;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfflineBiquadFilterNodeFakerProxy.prototype, "Q", {
        get: function () {
            // @todo Fake a proper AudioParam.
            var audioParam = {
                cancelScheduledValues: function (startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                defaultValue: 1,
                exponentialRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                linearRampToValueAtTime: function (value, endTime) {
                    endTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setTargetAtTime: function (target, startTime, timeConstant) {
                    startTime; // tslint:disable-line:no-unused-expression
                    target; // tslint:disable-line:no-unused-expression
                    timeConstant; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueAtTime: function (value, startTime) {
                    startTime; // tslint:disable-line:no-unused-expression
                    value; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                setValueCurveAtTime: function (values, startTime, duration) {
                    duration; // tslint:disable-line:no-unused-expression
                    startTime; // tslint:disable-line:no-unused-expression
                    values; // tslint:disable-line:no-unused-expression
                    return audioParam;
                },
                value: 1
            };
            return audioParam;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OfflineBiquadFilterNodeFakerProxy.prototype, "type", {
        get: function () {
            return this._type;
        },
        set: function (value) {
            this._type = value;
        },
        enumerable: true,
        configurable: true
    });
    OfflineBiquadFilterNodeFakerProxy.prototype.getFrequencyResponse = function (frequencyHz, magResponse, phaseResponse) {
        return this._nativeNode.getFrequencyResponse(frequencyHz, magResponse, phaseResponse);
    };
    return OfflineBiquadFilterNodeFakerProxy;
}(OfflineAudioNodeProxy));
export { OfflineBiquadFilterNodeFakerProxy };
var OfflineBiquadFilterNodeFaker = (function () {
    function OfflineBiquadFilterNodeFaker(_a) {
        var fakeNodeStore = _a.fakeNodeStore, nativeNode = _a.nativeNode, offlineAudioContext = _a.offlineAudioContext;
        this._node = null;
        this._proxy = new OfflineBiquadFilterNodeFakerProxy({ fakeNodeStore: fakeNodeStore, nativeNode: nativeNode, offlineAudioContext: offlineAudioContext });
        this._sources = new Map();
        fakeNodeStore.set(this._proxy, this);
    }
    Object.defineProperty(OfflineBiquadFilterNodeFaker.prototype, "proxy", {
        get: function () {
            return this._proxy;
        },
        enumerable: true,
        configurable: true
    });
    OfflineBiquadFilterNodeFaker.prototype.render = function (offlineAudioContext) {
        var _this = this;
        if (this._node !== null) {
            return Promise.resolve(this._node);
        }
        this._node = offlineAudioContext.createBiquadFilter();
        this._node.type = this._proxy.type;
        var promises = Array
            .from(this._sources)
            .map(function (_a) {
            var source = _a[0], _b = _a[1], input = _b.input, output = _b.output;
            // For some reason this currently needs to be a function body with a return statement. The shortcut syntax causes an error.
            return source
                .render(offlineAudioContext)
                .then(function (node) { return node.connect(_this._node, output, input); });
        });
        return Promise
            .all(promises)
            .then(function () { return _this._node; });
    };
    OfflineBiquadFilterNodeFaker.prototype.wire = function (source, output, input) {
        this._sources.set(source, { input: input, output: output });
        return this._proxy;
    };
    OfflineBiquadFilterNodeFaker.prototype.unwire = function (source) {
        this._sources.delete(source);
    };
    return OfflineBiquadFilterNodeFaker;
}());
export { OfflineBiquadFilterNodeFaker };
var OfflineBiquadFilterNodeFakerFactory = (function () {
    function OfflineBiquadFilterNodeFakerFactory() {
    }
    OfflineBiquadFilterNodeFakerFactory.prototype.create = function (_a) {
        var fakeNodeStore = _a.fakeNodeStore, nativeNode = _a.nativeNode, offlineAudioContext = _a.offlineAudioContext;
        return new OfflineBiquadFilterNodeFaker({
            fakeNodeStore: fakeNodeStore,
            nativeNode: nativeNode, offlineAudioContext: offlineAudioContext
        });
    };
    return OfflineBiquadFilterNodeFakerFactory;
}());
export { OfflineBiquadFilterNodeFakerFactory };
