// angular-audio-context/build/es2015/audio-context.module.js
import { CommonModule } from '@angular/common';
import { NgModule, OpaqueToken } from '@angular/core';
import { AudioContext as StandardizedAudioContextAudioContext, isSupported as standardizedAudioContextModuleIsSupported } from 'standardized-audio-context';
export const AudioContext = new OpaqueToken('AUDIO_CONTEXT_CONSTRUCTOR'); // tslint:disable-line:variable-name
// tslint:disable-line:variable-name
export function audioContextConstructorFactory() {
    return new StandardizedAudioContextAudioContext();
}
export const isSupported = new OpaqueToken('IS_SUPPORTED_PROMISE');
export function isSupportedFactory() {
    return standardizedAudioContextModuleIsSupported;
}
export class AudioContextModule {
}
AudioContextModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                providers: [
                    { provide: AudioContext, useFactory: audioContextConstructorFactory },
                    { provide: isSupported, useFactory: isSupportedFactory }
                ]
            },] },
];
/** @nocollapse */
AudioContextModule.ctorParameters = () => [];
